RecoTracker/TkSeedingLayers/src/SeedingLayerSetsBuilder.cc

```diff
@@ -363,12 +365,62 @@ std::vector<SeedingLayerSetsBuilder::SeedingLayerId> SeedingLayerSetsBuilder::la
 
 std::unique_ptr<SeedingLayerSetsHits> SeedingLayerSetsBuilder::hits(const edm::Event& ev, const edm::EventSetup& es) {
   updateEventSetup(es);
-
+ 
+  //-----------------new for FastSim-------------------------
+  edm::Handle<FastTrackerRecHitCollection> fastSimrechits_;
+  ev.getByToken(fastSimrecHitsToken_,fastSimrechits_);
+  edm::ESHandle<TrackerTopology> trackerTopology;
+  es.get<TrackerTopologyRcd>().get(trackerTopology);
+  SeedingLayerSetsHits::OwnedHits layerhits_;
+  //---------------------------------------------------------
+  
   auto ret = std::make_unique<SeedingLayerSetsHits>(theNumberOfLayersInSet,
                                                     &theLayerSetIndices,
                                                     &theLayerNames,
                                                     &theLayerDets);
 
+  if(isFastSim){
+    for(auto& layer: theLayers) {
+      layerhits_.clear();
+      for(auto &rh : *fastSimrechits_){
+	std::string hitlayer;
+	int layerNo;
+	std::string side;
+	if( (rh.det()->geographicalId()).subdetId() == PixelSubdetector::PixelBarrel){
+	  layerNo = (*trackerTopology.product()).pxbLayer(rh.det()->geographicalId());
+	  if(layerNo == 1)
+	    hitlayer = "BPix1";
+	  if(layerNo == 2)
+	    hitlayer = "BPix2";
+	  if(layerNo == 3)
+	    hitlayer = "BPix3";
+	  if(layerNo == 4)
+	    hitlayer = "BPix4";
+	}
+	else if ((rh.det()->geographicalId()).subdetId() == PixelSubdetector::PixelEndcap){
+	  layerNo = (*trackerTopology.product()).pxfDisk(rh.det()->geographicalId());
+	  side = (*trackerTopology.product()).pxfSide(rh.det()->geographicalId())==1 ? "_neg" : "_pos";
+	  if(layerNo == 1)
+	    hitlayer = "FPix1"+side;
+	  if(layerNo == 2)
+	    hitlayer = "FPix2"+side;
+	  if(layerNo == 3)
+	    hitlayer = "FPix3"+side;
+	}
+	else hitlayer = "UNKWN"; //strip                                                                                                                                    
+	if(theLayerNames[layer.nameIndex] == hitlayer){
```

Comment:
Let's try to avoid the string comparisons here as well. All the necessary information is already in the `layer`
* `layer.subdet` holds the subdetector
   * Unfortunately in the `GeomDetEnumerators::SubDetector` numbering scheme which is not bit-compatible with `PixelSubdetector`
* `layer.idLayer` has the layer/disk that can be directly compared to `pxbLayer()`/`pxfDisk()`
* `layer.side` has the FPix side that can be directly compared to `pxfSide()`

Reply:
Yes, taken care of this in the new FastSim-specific function for SeedingLayerSetsHits.